### ISSUE TITLE
VSCode "exports all" error confused with "imports all" for quite specific case?

### ISSUE THREAD

**Issue by leouk** [overturetool/vdm-vscode#174]
Hi Nick,

I was getting some weird behaviour on VSCode (Which is not in VDMJ) for this kind of spec:

```
module TestV2IState 
exports all
definitions 

state S of  
   x : nat 
   y : real
inv mk_S(x,y) == x > 15 and y < 5
init s == s = mk_S(25,0)
end

end TestV2IState
```

I get errors about imports all?

```
[{
	"resource": "/Users/nljsf/Local/reps/git/VDM_Toolkit/plugins/vdm2isa/src/test/resources/TestV2IState.vdmsl",
	"owner": "vdm-vscode",
	"code": "3190",
	"severity": 8,
	"message": "Import all from module with no exports?",
	"startLineNumber": 1,
	"startColumn": 8,
	"endLineNumber": 1,
	"endColumn": 9
}]
```

I think it happened on any spec that didn't have "anything" to export. If I add a 

```
types T = nat;
```

the error goes away. But if I remove the exports all (given my use of -strict), it fails. 

This error doesn't happen on VDMJ command line (lastest from development).

Best,
Leo

**Comment by nickbattle** [comment]
Hmm. I can only see this error if I combine your spec with another one that does `imports from TestV2IState all`. But then the error appears in command line VDMJ as well.

Can you create a complete example that fails in VSCode but not in VDMJ?

You're right about the root cause here. It's saying that you're asking to import everything, but actually the module has no definitions to export (as opposed to no exports definition!).  Perhaps this is confusing. I've tweaked it so that the two cases have different wording:
```
Error 3190: Import all from module with no exported definitions? in 'A' (test2.vdmsl) at line 2:27
Error 3190: Import all from module with no exports clause? in 'A' (test2.vdmsl) at line 2:27
```
Is that better?

**Comment by nickbattle** [comment]
Ugh, I can't remember the rules about whether `exports all` exports the type definition of the state record. I know it doesn't export the state fields or the "S" record value of that type.

You should be able to add the "target" folder to `.vscode/vdmignore` rather than using an ordering file. That's a bit easier to maintain and closer to what you really mean?

I can't see how an ordering file would affect this though. Strange. But unless I can reproduce it, I'm stuck :-(

**Comment by nickbattle** [comment]
PS. No, it looks like all state-related definitions are excluded from the export. So everything about state is local to the module defining it. But I do remember everything about import/export being rather poorly defined, since it was not in place when the ISO standard work was done. I tried to make VDMJ the same as VDMTools.

**Comment by leouk** [comment]
Okay, here is a very peculiar example - it does stem from a real one though! 

```
module TestV2IState 
exports 
   types 
      struct S
definitions 

state S of  
   x : nat 
   y : real
inv mk_S(x,y) == x > 15 and y < 5
init s == s = mk_S(25,0)
end

end TestV2IState

module TestV2IStateNoInit
imports from TestV2IState 
   types S renamed S
exports all
definitions 

state T of  
   x : nat 
   y : real
inv mk_T(x, y) == exists s in set {mk_S(25, 0)} & s.x = x and s.y = y

end

end TestV2IStateNoInit
```

**Comment by nickbattle** [comment]
The state record type "S" is excluded from an `exports all`, but you are allowed to explicitly type export it. I suppose you could use it to pass data to operations in the module to update the state. VDMTools is happy with this too.

So it looks like there's a bug somewhere when it looks up the type S in your example. I'll take a look :-)

**Comment by nickbattle** [comment]
OK, that's now fixed (development). It was because the rename wasn't correctly preserving the fact that this was a type (because it's also a state definition, which is unusual). I had to add an initializer to the 2nd module to get it to work:
```
init t == t = mk_T(25,0)
```
Otherwise it fails when comparing the undefined field values.

**Comment by leouk** [comment]
This gives an error on mk_S, but the exports of S as type is fine(!). I think the real example was something associated with "refinement" simulations through the interpretation of invariants.

I think in the experiments GC or Heap on VDM_Toolkit. Going there I can see I am not quite importing the state as a type. So maybe indeed there is nothing being exports.

In any case, the "access" to S as a type is funny, given it seems it can't be used anywhere?

**Comment by nickbattle** [comment]
I agree it's a very strange thing to be doing!  But now it's fixed, I think you could use it to pass values in a record. Not that this is very useful (and it's confusing, so I wouldn't recommend it style wise).

**Comment by leouk** [comment]
> I agree it's a very strange thing to be doing! But now it's fixed, I think you could use it to pass values in a record. Not that this is very useful (and it's confusing, so I wouldn't recommend it style wise).

I now realised that in the examples I had in mind I was using S1 S2 etc as actual record types, then doing the refinement as state (i.e. if you see VDM_Toolkit/experiments/vdm/GC or HEAP). 

In any case, the "spurious" example was nice to highlight this :-). 

Many thanks Nick.

**Comment by nickbattle** [comment]
I think this can be closed. The VDMJ fix will now be in 1.3.7.
